### ISSUE TITLE
HTML tokenizer consumes comments

### ISSUE THREAD

**Issue by labdsf** [jgm/pandoc#4282]
I just found (by running round-trip test in a loop) that Muse reader fails to parse `<code>` tag in this case:
```
$ pandoc -f muse -t native
<code><?foo</code>
[Para [Str "<code><?foo</code>"]]
```

HTML reader has similar problems:
```
$ printf '<code><?foo</code>' | pandoc -f html -t native
[]
```

But it is ok in this case:
```
$ printf '<code>a<?foo</code>' | pandoc -f muse -t native
[Para [Code ("",[],[]) "a<?foo"]]
```

HTML reader still fails:
```
$ printf '<code>a<?foo</code>' | pandoc -f html -t native
[Plain [Str "a"]]
```

Same problems are with "<!".

Related bug: #4094

**Comment by jgm** [comment]
The HTML reader starts by tokenizing using tagsoup, which is supposed to use official HTML5 tokenization rules.  Here's what we get:
```
Prelude Text.HTML.TagSoup> parseTags "<code><?foo</code>"
[TagOpen "code" [],TagOpen "?foo<" [("code","")]]
Prelude Text.HTML.TagSoup> parseTags "<code>a<?foo</code>"
[TagOpen "code" [],TagText "a",TagOpen "?foo<" [("code","")]]
```
If this tokenization is correct, then I'm not sure I see a problem with pandoc's output.

**Comment by jgm** [comment]
Note that wiki engines often have ad hoc ways of parsing `<code>` tags that don't line up with HTML's.

**Comment by labdsf** [comment]
Now I understand the problem better. HTML parsing is correct. But then, why did `htmlTag` consume "bogus comment" when I asked to match open tag for "code"?

```
$ pandoc -f muse -t native
<code><?bogus comment>bar</code>
[Para [Code ("",[],[]) "bar"]]
```

It also consumes normal comments:
```
$ pandoc -f muse -t native
<code><!--foo-->bar</code>
[Para [Code ("",[],[]) "bar"]]
```

Here it does not get to tokenize comment and leaves it alone:
```
$ pandoc -f muse -t native
<code>foo<!--bar-->baz</code>
[Para [Code ("",[],[]) "foo<!--bar-->baz"]]
```

This problem affects markdown:
```
$ pandoc -f markdown -t native
<div><!--foo-->bar</div>
[Div ("",[],[])
 [Plain [Str "bar"]]]
```

MediaWiki:
```
$ pandoc -f mediawiki -t native
<nowiki><!--foo-->bar</nowiki>
[Para [Str "bar"]]
```

**Comment by jgm** [comment]
The reason is quite simple.  The pandoc Code element only contains a string: it cannot contain formatting or "raw HTML."  When you write
```
<code>foo<!--bar-->baz</code>
```
the `<!--bar-->` is an HTML comment, not part of the code that is being quoted. (It won't show up when you view this with a browser, for example.)  So, parsing it as `Code ("",[],[]) "foo<!--bar-->baz"` would be incorrect.  Since we can't insert an HTML comment into the code (a plain string), we just leave it out.

So, I don't see any problem with pandoc's HTML reader here.  As for Muse, I don't know how muse interpreters are supposed to work, but if they're supposed to work as you indicate, then you won't be able to use functions exported by the HTML reader for code tags.

I'm also not sure about the proper behavior for `nowiki` nesting in mediawiki.

The Markdown example is strange.  The comment should be passed through as raw HTML, I'm not sure why it isn't, but I don't think this has anything to do with what's happening in the code example.

**Comment by jgm** [comment]
Here's a good contrast bringing out the problem in markdown:
```
% pandoc -t native
<div><!--foo-->bar</div>
[Div ("",[],[])
 [Plain [Str "bar"]]]
% pandoc -t native
<div> <!--foo-->bar</div>
[Div ("",[],[])
 [Plain [RawInline (Format "html") "<!--foo-->",Str "bar"]]]
```
Just adding the space before the comment makes it appear.

**Comment by jgm** [comment]
Another contrast:
```
% pandoc -t native
<div><!--foo-->bar</div>
[Div ("",[],[])
 [Plain [Str "bar"]]]
% pandoc -t native
<section><!--foo-->bar</section>
[RawBlock (Format "html") "<section><!--foo-->"
,Plain [Str "bar"]
,RawBlock (Format "html") "</section>"]
```

**Comment by jgm** [comment]
OK, you're right, it appears that `htmlTag (~== TagOpen "div" [])` does consume `<div><!--foo-->` in this example.

**Comment by jgm** [comment]
I suspect another tagsoup bug:
```
Prelude Text.HTML.TagSoup> parseTagsOptions parseOptions{ optTagWarning = True, optTagPosition = True } "<div><!--foo-->bar</div>"
[TagPosition 1 1,TagOpen "div" [],TagPosition 1 8,TagComment "foo",TagPosition 1 16,TagText "bar",TagPosition 1 19,TagClose "div"]
```
Notice `TagPosition 1 8`.  That's wrong.  At this point we've only consumed `<div>`, so it should be `TagPosition 1 6`.  I'll open an issue on the tagsoup repo.

**Comment by jgm** [comment]
There should be a new release of tagsoup soon, and then we can fix this.

**Comment by labdsf** [comment]
Bug is fixed for normal comments (`<nowiki><!--foo--></nowiki>`), but not for bogus comments:
```
$ pandoc -f mediawiki -t native
<nowiki><?foo</nowiki>
^D
[]
```

**Comment by jgm** [comment]
`htmlTag` treats bogus comments specially -- see the code.  This is because things like `<http://google.com>`, which should be markdown autolinks, are parsed as bogus comments.

Can you determine whether there's a tagsoup bug (using tagsoup directly)?  I suspect this is simply the intended behavior of `htmlTag`, which may not fit your needs perfectly.

**Comment by labdsf** [comment]
It is all the same bug with incorrect TagPosition (note the difference: `TagPosition 1 6` vs `TagPosition 1 8`):
```
> parseTagsOptions Text.HTML.TagSoup.parseOptions{ optTagWarning = True, optTagPosition = True } "<div><!--foo-->bar</div>"
[TagPosition 1 1,TagOpen "div" [],TagPosition 1 6,TagComment "foo",TagPosition 1 16,TagText "bar",TagPosition 1 19,TagClose "div"]
> parseTagsOptions Text.HTML.TagSoup.parseOptions{ optTagWarning = True, optTagPosition = True } "<div><?foo</div>"
[TagPosition 1 1,TagOpen "div" [],TagPosition 1 8,TagOpen "?foo<" [("div","")],TagPosition 1 13,TagWarning "Unexpected \"/\"",TagPosition 1 17,TagWarning "Expected \"?>\""]
```

**Comment by labdsf** [comment]
I opened an issue for tagsoup: https://github.com/ndmitchell/tagsoup/issues/71

**Comment by labdsf** [comment]
One more similar bug:
```
$ pandoc -f mediawiki -t native
<nowiki><?foo</nowiki>
[Para [Str "<?foo"]]
$ pandoc -f mediawiki -t native
<nowiki><?</nowiki>
[]
```

**Comment by ndmitchell** [comment]
I just released tagsoup-0.14.5 which I think will fix the bug that you would boil down from the test case above. Let me know if you become aware of any others. Apologies for the drip-drip feed of bugs/fixes, I am adding them to my test suite as I come across them, but it feels like you've managed to determine an algebraic property that I haven't!

**Comment by labdsf** [comment]
One more example wiki 0.14.5: `<nowiki><!</nowiki>`

**Comment by ndmitchell** [comment]
Thanks - looking at it. How easy is it for you to test against a version of tagsoup in git?

**Comment by ndmitchell** [comment]
I've just pushed some changes which might help, although `<!` is a bit special because it starts a comment (according to the HTML 5 spec).

**Comment by ndmitchell** [comment]
I've released the changes as 0.14.6.

**Comment by labdsf** [comment]
@ndmitchell Thanks, now round-trip test finds (very rare) bugs not related to tagsoup.
